jsign-core/src/main/java/net/jsign/mscab/MSCabFile.java

```diff
@@ -0,0 +1,423 @@
+/**
+ * Copyright 2019 Emmanuel Bourg
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.jsign.mscab;
+
+import net.jsign.DigestAlgorithm;
+import net.jsign.Signable;
+import net.jsign.asn1.authenticode.AuthenticodeObjectIdentifiers;
+import net.jsign.asn1.authenticode.SpcAttributeTypeAndOptionalValue;
+import net.jsign.asn1.authenticode.SpcIndirectDataContent;
+import net.jsign.asn1.authenticode.SpcPeImageData;
+import org.bouncycastle.asn1.ASN1Encodable;
+import org.bouncycastle.asn1.ASN1InputStream;
+import org.bouncycastle.asn1.ASN1Object;
+import org.bouncycastle.asn1.DERNull;
+import org.bouncycastle.asn1.cms.Attribute;
+import org.bouncycastle.asn1.cms.AttributeTable;
+import org.bouncycastle.asn1.cms.ContentInfo;
+import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
+import org.bouncycastle.asn1.x509.DigestInfo;
+import org.bouncycastle.cms.CMSException;
+import org.bouncycastle.cms.CMSProcessable;
+import org.bouncycastle.cms.CMSSignedData;
+import org.bouncycastle.cms.SignerInformation;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * MSCabinet File.
+ *
+ * This class is thread safe.
+ *
+ * @see <a href="http://download.microsoft.com/download/5/0/1/501ED102-E53F-4CE0-AA6B-B0F93629DDC6/Exchange/[MS-CAB].pdf">[MS-CAB]: Cabinet File Format</a>
+ *
+ * @author Joseph Lee
+ * @since 3.2
+ */
+public class MSCabFile implements Signable, Closeable {
+    private final CFHeader header = new CFHeader();
+    private int sigpos = 0;
+    private int siglen = 0;
+
+    private File file;
+    final SeekableByteChannel channel;
+
+    /** Reusable buffer for reading bytes, words, dwords and qwords from the file */
+    private final ByteBuffer valueBuffer = ByteBuffer.allocate(8);
+    {
+        valueBuffer.order(ByteOrder.LITTLE_ENDIAN);
+    }
+
+    /**
+     * Tells if the specified file is a Portable Executable file.
```

Comment:
The javadoc can be updated here to replace `Portable Executable` with `MS Cabinet`